### PR TITLE
Support Python 3.11

### LINKED ISSUE #5911: Incompatibility with python 3.11
**Describe the bug**

There are several places in the rdkit code where we use the construct `random.shuffle(v, random=random.random)` to shuffle the contents of lists. The `random` keyword argument to `shuffle()` has been deprecated since python 3.9 and was removed in python 3.11: https://docs.python.org/3.11/library/random.html#random.shuffle

According to the documentation (https://docs.python.org/3.10/library/random.html#random.shuffle) if you don't provide the argument `random.random()` will be used, so you should get the same results. You most definitely do not:
```
$ python -c 'import random;random.seed(0xf00d);l=list(range(10));random.shuffle(l);print(l);l=list(range(10));random.seed(0xf00d);random.shuffle(l,random=random.random);print(l)'
[8, 3, 0, 5, 6, 9, 1, 4, 2, 7]
<string>:1: DeprecationWarning: The *random* parameter to shuffle() has been deprecated
since Python 3.9 and will be removed in a subsequent version.
[9, 5, 7, 3, 0, 8, 6, 2, 1, 4]
```
Dealing with this is going to require changing the expected results for numerous tests. 

Oh, what fun.

**Configuration (please complete the following information):**
 - RDKit version: master
 - OS: all
 - Python version (if relevant): 3.11+

### PR DESCRIPTION
**Pull request by greglandrum**
This required minor changes to the way pickling is handled and a switch away from using `random.shuffle` with the `random` keyword.

Also adds a CI build for py 3.11

Fixes #5911

### CODE REVIEW

**Comment by greglandrum** [comment]
@bp-kelley : this is an alternate approach to handle the deprecation of the extra argument to `random.shuffle()` and probably supersedes #5929

**Review by ricrogz — COMMENTED** [review]
LGTM

**Comment by bp-kelley** [comment]
I’ll be able to take a look this weekend.  If it’s better Han the previous random fix, I’m all for it.

File: Code/RDBoost/Wrap.h [hunk 1]
```diff
@@ -306,4 +306,28 @@ struct iterable_converter {
...
+              .ptr());
+      python::throw_error_already_set();
+    }
+
+    // restore the object's __dict__
+    python::dict d = python::extract<python::dict>(w_obj.attr("__dict__"))();
+    d.update(state[0]);
+  }
+
+  static bool getstate_manages_dict() { return true; }
```
Comment (bp-kelley) [inline Code/RDBoost/Wrap.h:330]: This is a potentially breaking change.  Previously you could pickle arbitrary things attached to molecules:


```
>>> from rdkit import Chem
>>> m = Chem.MolFromSmiles("C")
>>> m.a = "foo"
>>> import pickle
>>> s = pickle.dumps(m)
>>> m2 = pickle.loads(s)
>>> m2.a
'foo'
```

Comment (greglandrum) [reply Code/RDBoost/Wrap.h:330]: > This is a potentially breaking change. Previously you could pickle arbitrary things attached to molecules:
> 
> ```
> >>> from rdkit import Chem
> >>> m = Chem.MolFromSmiles("C")
> >>> m.a = "foo"
> >>> import pickle
> >>> s = pickle.dumps(m)
> >>> m2 = pickle.loads(s)
> >>> m2.a
> 'foo'
> ```

@bp-kelley and I talked about this on slack. I don't think anything breaks, in fact there's a test for this case in the PR

File: rdkit/RDRandom.py [hunk 2]
```diff
@@ -18,3 +17,30 @@
 random = _random.random
 randrange = _random.randrange
 seed = _random.seed
+
+if sys.hexversion >= 0x30b0000:
```
Comment (bp-kelley) [inline rdkit/RDRandom.py:21]: Nice fix.

**Comment by bp-kelley** [comment]
LGTM

**Pull request merged**